cpp/include/patterns/update_frontier_v_push_if_out_nbr.cuh

```diff
@@ -515,49 +522,74 @@ void update_frontier_v_push_if_out_nbr(
-    thrust::upper_bound(d_vertex_lasts.begin(),
-                        d_vertex_lasts.end(),
+    thrust::lower_bound(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
                         buffer_key_first,
                         buffer_key_first + num_buffer_elements,
+                        d_vertex_lasts.begin(),
+                        d_vertex_lasts.end(),
                         d_tx_buffer_last_boundaries.begin());
     std::vector<edge_t> h_tx_buffer_last_boundaries(d_tx_buffer_last_boundaries.size());
     raft::update_host(h_tx_buffer_last_boundaries.data(),
                       d_tx_buffer_last_boundaries.data(),
                       d_tx_buffer_last_boundaries.size(),
                       handle.get_stream());
-    std::vector<edge_t> tx_counts(h_tx_buffer_last_boundaries.size());
+    CUDA_TRY(cudaStreamSynchronize(handle.get_stream()));
+    std::vector<size_t> tx_counts(h_tx_buffer_last_boundaries.size());
     std::adjacent_difference(
       h_tx_buffer_last_boundaries.begin(), h_tx_buffer_last_boundaries.end(), tx_counts.begin());
 
-    std::vector<edge_t> rx_counts(graph_view.is_hypergraph_partitioned() ? row_comm_size
+    std::vector<size_t> rx_counts(graph_view.is_hypergraph_partitioned() ? row_comm_size
                                                                          : col_comm_size);
     std::vector<raft::comms::request_t> count_requests(tx_counts.size() + rx_counts.size());
+    size_t tx_self_i = std::numeric_limits<size_t>::max();
     for (size_t i = 0; i < tx_counts.size(); ++i) {
-      comm.isend(&tx_counts[i],
-                 1,
-                 graph_view.is_hypergraph_partitioned() ? col_comm_rank * row_comm_size + i
-                                                        : row_comm_rank * col_comm_size + i,
-                 0 /* tag */,
-                 count_requests.data() + i);
+      auto comm_dst_rank = graph_view.is_hypergraph_partitioned()
+                             ? col_comm_rank * row_comm_size + static_cast<int>(i)
+                             : row_comm_rank * col_comm_size + static_cast<int>(i);
+      if (comm_dst_rank == comm_rank) {
+        tx_self_i = i;
+        // FIXME: better define request_null (similar to MPI_REQUEST_NULL) under raft::comms
+        count_requests[i] = std::numeric_limits<raft::comms::request_t>::max();
+      } else {
+        comm.isend(&tx_counts[i], 1, comm_dst_rank, 0 /* tag */, count_requests.data() + i);
+      }
     }
     for (size_t i = 0; i < rx_counts.size(); ++i) {
-      comm.irecv(&rx_counts[i],
-                 1,
-                 graph_view.is_hypergraph_partitioned() ? col_comm_rank * row_comm_size + i
-                                                        : row_comm_rank + i * row_comm_size,
-                 0 /* tag */,
-                 count_requests.data() + tx_counts.size() + i);
+      auto comm_src_rank = graph_view.is_hypergraph_partitioned()
```

Comment:
Given the large number of `graph_view.is_hypergraph_partitioned()` tests (many inside loops), perhaps might be better to use a Template Method / Strategy design pattern here, where different strategies are created depending on whether the (hyper)graph is partitioned or not. And then pass the strategy as an argument to a function that abstracts away these decisions. This might optimize away such tests, but more importantly it would make the code more readable.

Example:

Instead of:
```
foo(graph_view_t const& g, args...)
{
   for(auto i =0;...,++i)
   {
     auto comm_dst_rank = graph_view.is_hypergraph_partitioned() ? ... : ...;
     do_something(comm_dst_rank, args...);  
   }
}

//client call:
graph_view_t g_view{...};
//...
foo(g_view, args...);
```
Consider:
```
struct graph_partitioning_strategy_t {
  graph_partitioning_strategy_t(graph_view_t const& g_view, size_t row_comm_size, size_t coll_comm_size, size_t comm_rank):
     comm_dst_rank_start_(g_view.is_hypergraph_partitioned()? col_comm_rank * row_comm_size : comm_rank / col_comm_size)
  {...}
  
  int comm_dst_rank(int i) const { return comm_dst_rank_start_ + i; }
private:
   int comm_dst_rank_start_;
};

foo(graph_partitioning_strategy_t const& part_gs, args...)
{
   for(auto i =0;...,++i)
   {
     do_something(part_gs.comm_dst_rank(), args...);  
   }
}

//client call:
graph_view_t g_view{...};
graph_partitioning_strategy_t part_gs(g_view, ...);
//...

//...
foo(part_gs, args...);
```